ops/model.py

```diff
@@ -797,6 +927,267 @@ def __init__(self, name: str, address_info: '_AddressDict'):
+                'cannot grant secret access to application that is not '
+                'remote to the specified relation')
+
+        relation = relation or target
+        assert isinstance(relation, Relation)  # type guard
+        unit = target.name if isinstance(target, Unit) else None
+        self._backend.secret_grant(self.id, relation.id, unit_id=unit)
+
+    def revoke(self, target: Union['Relation', Unit, Application],
+               relation: 'Relation' = None):
+        """Revoke read access to this secret from a designated target.
+        Parameters
+        ----------
+        target: The individual unit, application, or relation to revoke secret read access from.
+        relation: The relation that this secret is bound to.
+            Needs to be provided if `target` is not a Relation instance.
+        """
+
+        if not self._am_owner:
+            raise RuntimeError(
+                'cannot revoke secret {} which is owned by a different unit'.format(self))
+        if relation is None and not isinstance(target, Relation):
+            raise RuntimeError(
+                'revoking secret access to a unit or application target requires a relation to be specified')
+        if isinstance(target, Application) and relation.app != target:
+            raise RuntimeError(
+                'cannot revoke secret access from application that is not remote to the specified relation')
+
+        relation = relation or target
+        assert isinstance(relation, Relation)  # type guard
+
+        unit = target.name if isinstance(target, Unit) else None
+        self._backend.secret_revoke(self.id, relation.id, unit_id=unit)
+
+    def prune(self):
+        """Remove this secret revision from juju.
+
+        This should be called while handling `secret-remove` events, otherwise
+        you cannot be ensured that this revision is not still being consumed by
+        some remote units.
+        """
+        if not self._am_owner:
+            raise RuntimeError(
+                'cannot prune secret {} which is owned by a different unit'.format(self))
+        if self._revision is None:
+            raise RuntimeError(
+                'secret {} has no revision to prune - are you outside a secret-remove hook call?')
+        self._backend.secret_remove(self.id, revision=self._revision)
+
+    def remove(self):
+        """Remove all revisions of this secret from juju, effectively deleting it.
+
+        This should typically happen when the relation the secret is bound to is broken.
+        Removing the secret prematurely means that all units/applications it has
+        been granted to might attempt to get the secret
+        """
+
+        if not self._am_owner:
+            raise RuntimeError(  # todo unify with SecretOwnershipError
+                'cannot remove secret {} which is owned by a different unit'.format(self))
+        self._backend.secret_remove(self.id)
+
+    def update(self) -> '_Secret':
```

Comment:
The issue with `get_` is that it may give the wrong impression that no changes are being made to the model, that we are just inspecting another revision. So we opted for a word that was a bit more 'active' than 'get'.

`update` is in fact as you say going to update the secret in juju.

```
secret = self.model.get_secret(id=X)
new_secret = secret.update()
```

is basically syntactic sugar for the following:

```
secret = self.model.get_secret(id=X)
secret.get('foo', update=True)
new_secret = self.model.get_secret(id=X)
```

We thought it would be a common use case, but if it's confusing we can also remove it.

Reply:
The secret is NOT being updated. The revision which is being tracked is. I agree with Ben that calling this `update()` is misleading, especially as it is invoked as `secret.update()`. The CLI is done as `secret-get <id> --update` so the context is a lot more unambiguous. We really are talking about an operation akin to "get latest revision and from now on track that one". So maybe something like "get_latest_revision()" and "peek_latest_revision()" would make more sense.

Reply:
I renamed update to get_latest_revision, but I realized it feels a bit awkward to also have update as an optional kwarg to `get()`. 
I think we should go one of two ways:
- get rid of get_latest_revision and use 'get()` only, with optional `update` and `peek` kwargs. Staying close to the cli.
- remove all optional kwargs from `get()` and add a `peek_latest_revision` method which returns a Secret object that's locked in 'peek mode': its `.get()` method will behave as currently `.get(peek=True)` does.

WDYT?

Reply:
The first option above seems good to me (stay close to cli). Just IMO. 

But we could also consider:
`get(track_latest=False)`
`peek()`

Or something like that.